File: solutions/1381. Design a Stack With Increment Operation/1381.cpp
// Time complexity: O(1)
// Space complexity: O(n)
class CustomStack {
 public:
 CustomStack(int maxSize) {
  stackArray.resize(maxSize);
  incrementArray.resize(maxSize);
  topIndex = -1;
 }

 void push(int x) {
  if (topIndex < (int)(stackArray.size()) - 1)
    stackArray[++topIndex] = x;
 }

 int pop() {
  if (topIndex < 0)
    return -1;

  int result = stackArray[topIndex] + incrementArray[topIndex];

  if (topIndex > 0)
    incrementArray[topIndex - 1] += incrementArray[topIndex];

  incrementArray[topIndex] = 0;
  --topIndex;

  return result;
 }

 void increment(int k, int val) {
  if (topIndex >= 0) {
    int index = min(topIndex, k - 1);
    incrementArray[index] += val;
  }
 }

 private:
  vector<int> stackArray;
  vector<int> incrementArray;
  int topIndex;
};

File: solutions/1381. Design a Stack With Increment Operation/1381.rs
// Time complexity: O(1)
// Space complexity: O(n)
struct CustomStack {
    stack_array: Vec<i32>,
    increment_array: Vec<i32>,
    top_index: isize,
}

impl CustomStack {
    fn new(max_size: i32) -> Self {
        CustomStack {
            stack_array: vec![0; max_size as usize],
            increment_array: vec![0; max_size as usize],
            top_index: -1,
        }
    }

    fn push(&mut self, x: i32) {
        if self.top_index < (self.stack_array.len() as isize) - 1 {
            self.top_index += 1;
            self.stack_array[self.top_index as usize] = x;
        }
    }

    fn pop(&mut self) -> i32 {
        if self.top_index < 0 {
            return -1;
        }
        let idx = self.top_index as usize;
        let result = self.stack_array[idx] + self.increment_array[idx];

        if self.top_index > 0 {
            self.increment_array[idx - 1] += self.increment_array[idx];
        }
        self.increment_array[idx] = 0;
        self.top_index -= 1;

        result
    }

    fn increment(&mut self, k: i32, val: i32) {
        if self.top_index >= 0 {
            let increment_index = (self.top_index as usize).min(k as usize - 1);
            self.increment_array[increment_index] += val;
        }
    }
}
